tests/python_package_test/test_engine.py

```diff
@@ -2887,3 +2888,30 @@ def hook(obj):
     dumped_model_str = str(bst.dump_model(5, 0, object_hook=hook))
     assert "leaf_value" not in dumped_model_str
     assert "LV" in dumped_model_str
+
+
+def test_force_split_with_feature_fraction():
+    X, y = load_boston(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    lgb_train = lgb.Dataset(X_train, y_train)
+
+    forced_split = {
+        "feature": 0,
+        "threshold": 0.5
+    }
+
+    with open("forced_split.json", "w") as f:
+        f.write(json.dumps(forced_split))
+
+    params = {
+        "objective": "regression",
+        "feature_fraction": 0.6,
+        "force_col_wise": True,
+        "feature_fraction_seed": 1,
+        "forcedsplits_filename": "forced_split.json"
+    }
+
+    gbm = lgb.train(params, lgb_train)
```

Comment:
oooo interesting!! I didn't consider that fact that the choice of col-wise vs. row-wise histogram construction could be a source of randomness for small datasets (because the timing differences are so small). Thanks very much for explaining that.

But this PR is intended to FIX #4601, regardless of whether or not `force_col_wise=True`, right? I think that's the ideal behavior, since the setting of `force_col_wise` should just be an optimization and not impact the correctness of results or risk of errors.

If it is, then I think the unit tests should be testing once with `force_col_wise=True` and once with `force_row_wise=True`.

Reply:
I find another very random issue related to force_row_wise=True.... Good news is now I can repro it with this test if I change force-row-wise to True with seed 1 (it is not related to feature fraction, and looks like also not related to force split...), I believe it caused by some issues in row wise histogram construction, but I still not able to find the root cause. Let me work on that separately and add tests for row-wise in that fix.

Reply:
It looks like the possible root cause of #4739 and #3679